File: studentapp/lib/services/grade_service.dart
import '../models/models.dart';
import 'api_service.dart';

/// Service for handling grade-related API calls
class GradeService {
  static const String _endpoint = '/grades';

  /// Get all grades
  static Future<List<Grade>> getAllGrades() async {
    try {
      final response = await ApiService.get(_endpoint);

      // Handle array response
      if (response is List) {
        return response
            .whereType<Map<String, dynamic>>()
            .map((json) => Grade.fromJson(json))
            .toList();
      }

      // Handle object with data field
      if (response is Map<String, dynamic> && response['data'] is List) {
        final List data = response['data'];
        return data
            .whereType<Map<String, dynamic>>()
            .map((json) => Grade.fromJson(json))
            .toList();
      }

      return [];
    } catch (e) {
      throw Exception('Không thể tải danh sách điểm: $e');
    }
  }

  /// Get grade by ID
  static Future<Grade> getGradeById(int id) async {
    try {
      final response = await ApiService.get('$_endpoint/$id');
      return Grade.fromJson(response);
    } catch (e) {
      throw Exception('Không thể tải thông tin điểm: $e');
    }
  }

  /// Create new grade
  static Future<Grade> createGrade(Grade grade) async {
    try {
      final response = await ApiService.post(_endpoint, grade.toJson());
      return Grade.fromJson(response);
    } catch (e) {
      throw Exception('Không thể tạo điểm mới: $e');
    }
  }

  /// Update existing grade
  static Future<Grade> updateGrade(int id, Grade grade) async {
    try {
      final response = await ApiService.put('$_endpoint/$id', grade.toJson());
      return Grade.fromJson(response);
    } catch (e) {
      throw Exception('Không thể cập nhật điểm: $e');
    }
  }

  /// Delete grade
  static Future<void> deleteGrade(int id) async {
    try {
      await ApiService.delete('$_endpoint/$id');
    } catch (e) {
      throw Exception('Không thể xóa điểm: $e');
    }
  }

  /// Get grades by student ID
  static Future<List<Grade>> getGradesByStudentId(String studentId) async {
    try {
      final response = await ApiService.get('$_endpoint/student/$studentId');

      List<dynamic> gradesJson;
      if (response is List) {
        gradesJson = response;
      } else if (response.containsKey('data') && response['data'] is List) {
        gradesJson = response['data'];
      } else {
        gradesJson = [];
      }

      return gradesJson
          .whereType<Map<String, dynamic>>()
          .map((json) => Grade.fromJson(json))
          .toList();
    } catch (e) {
      throw Exception('Không thể tải điểm của sinh viên: $e');
    }
  }

  /// Get grades by subject ID
  static Future<List<Grade>> getGradesBySubjectId(String subjectId) async {
    try {
      final response = await ApiService.get('$_endpoint/subject/$subjectId');

      List<dynamic> gradesJson;
      if (response is List) {
        gradesJson = response;
      } else if (response.containsKey('data') && response['data'] is List) {
        gradesJson = response['data'];
      } else {
        gradesJson = [];
      }

      return gradesJson
          .whereType<Map<String, dynamic>>()
          .map((json) => Grade.fromJson(json))
          .toList();
    } catch (e) {
      throw Exception('Không thể tải điểm của môn học: $e');
    }
  }

  /// Get average score by student ID
  static Future<double> getAverageScoreByStudentId(String studentId) async {
    try {
      final response = await ApiService.get(
        '$_endpoint/student/$studentId/average',
      );
      return (response as num).toDouble();
    } catch (e) {
      throw Exception('Không thể tải điểm trung bình của sinh viên: $e');
    }
  }

  /// Get average score by subject ID
  static Future<double> getAverageScoreBySubjectId(String subjectId) async {
    try {
      final response = await ApiService.get(
        '$_endpoint/subject/$subjectId/average',
      );
      return (response as num).toDouble();
    } catch (e) {
      throw Exception('Không thể tải điểm trung bình của môn học: $e');
    }
  }
}

File: student-backend/src/main/java/com/company/student_backend/service/GradeService.java
package com.company.student_backend.service;

import com.company.student_backend.dto.GradeDTO;
import com.company.student_backend.exception.DuplicateResourceException;
import com.company.student_backend.exception.ResourceNotFoundException;
import com.company.student_backend.model.Grade;
import com.company.student_backend.model.Student;
import com.company.student_backend.model.Subject;
import com.company.student_backend.repository.GradeRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class GradeService {

    private final GradeRepository gradeRepository;
    private final ValidationService validationService;

    public List<GradeDTO> getAllGrades() {
        log.debug("Fetching all grades");
        return gradeRepository.findAllWithStudentAndSubject().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public GradeDTO getGradeById(Long id) {
        log.debug("Fetching grade with ID: {}", id);
        Grade grade = gradeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Không tìm thấy điểm với ID: " + id));
        return convertToDTO(grade);
    }

    public GradeDTO createGrade(GradeDTO gradeDTO) {
        log.debug("Creating new grade: {}", gradeDTO);

        // Validate student and subject exist using ValidationService
        Student student = validationService.validateAndGetStudent(gradeDTO.getStudentId());
        Subject subject = validationService.validateAndGetSubject(gradeDTO.getSubjectId());

        // Check if grade already exists for this student-subject combination
        if (gradeRepository.existsByStudentStudentIdAndSubjectSubjectId(gradeDTO.getStudentId(), gradeDTO.getSubjectId())) {
            throw new DuplicateResourceException("Điểm đã tồn tại cho học sinh " + gradeDTO.getStudentId() + " và môn học " + gradeDTO.getSubjectId());
        }

        Grade grade = new Grade();
        grade.setStudent(student);
        grade.setSubject(subject);
        grade.setAverageScore(gradeDTO.getAverageScore());

        Grade savedGrade = gradeRepository.save(grade);
        log.info("Created grade with ID: {}", savedGrade.getId());

        return convertToDTO(savedGrade);
    }

    public GradeDTO updateGrade(Long id, GradeDTO gradeDTO) {
        log.debug("Updating grade with ID: {}", id);

        Grade existingGrade = gradeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Không tìm thấy điểm với ID: " + id));

        existingGrade.setAverageScore(gradeDTO.getAverageScore());

        Grade updatedGrade = gradeRepository.save(existingGrade);
        log.info("Updated grade with ID: {}", updatedGrade.getId());

        return convertToDTO(updatedGrade);
    }

    public void deleteGrade(Long id) {
        log.debug("Deleting grade with ID: {}", id);

        if (!gradeRepository.existsById(id)) {
            throw new ResourceNotFoundException("Không tìm thấy điểm với ID: " + id);
        }

        gradeRepository.deleteById(id);
        log.info("Deleted grade with ID: {}", id);
    }

    public List<GradeDTO> getGradesByStudentId(String studentId) {
        log.debug("Fetching grades for student ID: {}", studentId);
        return gradeRepository.findByStudentStudentId(studentId).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public List<GradeDTO> getGradesBySubjectId(String subjectId) {
        log.debug("Fetching grades for subject ID: {}", subjectId);
        return gradeRepository.findBySubjectSubjectId(subjectId).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public BigDecimal getAverageScoreByStudentId(String studentId) {
        log.debug("Calculating average score for student ID: {}", studentId);
        return gradeRepository.findAverageScoreByStudentId(studentId);
    }

    public BigDecimal getAverageScoreBySubjectId(String subjectId) {
        log.debug("Calculating average score for subject ID: {}", subjectId);
        return gradeRepository.findAverageScoreBySubjectId(subjectId);
    }

    private GradeDTO convertToDTO(Grade grade) {
        GradeDTO dto = new GradeDTO();
        dto.setId(grade.getId());
        dto.setStudentId(grade.getStudent().getStudentId());
        dto.setSubjectId(grade.getSubject().getSubjectId());
        dto.setAverageScore(grade.getAverageScore());
        dto.setStudentName(grade.getStudent().getStudentName());
        dto.setSubjectName(grade.getSubject().getSubjectName());
        return dto;
    }
}
